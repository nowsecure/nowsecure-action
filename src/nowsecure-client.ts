/*
 * Copyright Â© 2021-2022 NowSecure Inc.
 *
 * SPDX-License-Identifier: MIT
 */

import * as http from "@actions/http-client";
import type {
  UploadApplicationResponse,
  PullReportResponse,
} from "./types/platform";
import { version } from "./nowsecure-version";
import { sleep } from "./utils";

export const USER_AGENT = `NowSecure GitHub Action/${version}`;
export const DEFAULT_API_URL = "https://api.nowsecure.com";
export const DEFAULT_LAB_API_URL = "https://lab-api.nowsecure.com";

/**
 * GraphQL request to check if baseline limit has been reached.
 *
 * NOTE: Must be wrapped in a `query` tag for bare requests!
 */
const LICENSE_GQL = `my {
  user {
    organization {
      usage {
        assessment {
          limit
          reached
          total
        }
      }
    }
  }
}`;

/**
 * GraphQL request for Platform.
 *
 * NOTE: Must be kept in sync with `PullReportResponse`.
 */
const platformGql = (reportId: string): string => `{
  ${LICENSE_GQL}
  auto {
    assessments(scope:"*" refs:["${reportId.replace(/[^0-9a-z-]/gi, "")}"]) {
      deputy: _raw(path: "yaap.complete.results[0].deputy.deputy.data[0].results")
      packageKey
      taskId
      applicationRef
      ref
      report {
        findings {
          kind
          key
          title
          summary
          affected
          severity
          uniqueVulnerabilityId
          context {
            fields
            rows
          }
          check {
            issue {
              title
              warn
              description
              impactSummary
              stepsToReproduce
              recommendation
              category
              cvss
              codeSamples {
                syntax
                caption
                block
              }
              guidanceLinks {
                caption
                url
              }
            }
          }
        }
      }
    }
  }
}`;

export class NowSecure {
  #client: http.HttpClient;
  #apiUrl: string;
  #labApiUrl: string;

  constructor(
    platformToken: string,
    apiUrl: string = DEFAULT_API_URL,
    labApiUrl: string = DEFAULT_LAB_API_URL
  ) {
    this.#apiUrl = apiUrl;
    this.#labApiUrl = labApiUrl;
    this.#client = new http.HttpClient(USER_AGENT, undefined, {
      allowRetries: true,
      maxRetries: 3,
      headers: {
        Authorization: `Bearer ${platformToken}`,
        Accept: "application/json",
        "Content-Type": "application/json",
      },
    });
  }

  /**
   * Pulls a report from NowSecure. Throws an exception if an error occurs.
   */
  async pullReport(reportId: string): Promise<PullReportResponse> {
    const r = await this.#client.getJson<PullReportResponse>(
      `${this.#apiUrl}/graphql?query=${platformGql(reportId)}`
    );

    if (r.statusCode !== 200) {
      throw new Error(`Report request failed with status ${r.statusCode}`);
    }

    return r.result;
  }

  /**
   * Poll Platform to resolve the report ID to a report.
   * GitHub Actions will handle the timeout for us in the event something goes awry.
   */
  async pollForReport(
    reportId: string,
    pollInterval: number
  ): Promise<PullReportResponse | null> {
    let report = null;
    for (;;) {
      console.log("Checking for NowSecure report... ", reportId);
      report = await this.pullReport(reportId);
      // NOTE: No optional chaining on Node.js 12 in GitHub Actions.
      try {
        if (report.data.auto.assessments[0].report) {
          break;
        } else {
          await sleep(pollInterval);
        }
      } catch (e) {
        console.error(e);
        // No report data.
      }
    }
    return report;
  }

  /**
   * Upload an application binary stream to NowSecure Platform and return job
   * details. Throws an exception if an error occurs.
   */
  async submitBin(
    stream: NodeJS.ReadableStream,
    groupId: string
  ): Promise<UploadApplicationResponse> {
    const r = await this.#client.sendStream(
      "POST",
      `${this.#labApiUrl}/build/?group=${groupId}`,
      stream,
      {}
    );

    if (r.message.statusCode !== 200) {
      throw new Error(
        `Application upload failed with status ${r.message.statusCode}`
      );
    }

    const body = await r.readBody();
    return JSON.parse(body);
  }

  /**
   * Checks if the assessment limit has been reached. Throws an exception if
   * an error occurs.
   */
  async isLicenseValid(licenseWorkaround: boolean): Promise<boolean> {
    const r = await this.#client.getJson<PullReportResponse>(
      `${this.#apiUrl}/graphql?query={${LICENSE_GQL}}`
    );

    if (r.statusCode !== 200) {
      throw new Error(`Report request failed with status ${r.statusCode}`);
    }

    const { total, limit, reached } =
      r.result.data.my.user.organization.usage.assessment;

    let limitReached = reached;
    if (licenseWorkaround) {
      // FIXME: Workaround platform license counting issue.
      limitReached = limit !== -1 && total + 1 >= limit;
    }

    return !limitReached;
  }
}
