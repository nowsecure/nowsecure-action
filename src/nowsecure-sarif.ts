/*
 * Copyright Â© 2021-2022 NowSecure Inc.
 *
 * SPDX-License-Identifier: MIT
 */

import type {
  Log,
  Run,
  ReportingDescriptor,
  Result,
  Notification,
  PhysicalLocation,
} from "sarif";
import type { PullReportResponse, Severity } from "./types/platform";
import { ripGrep as rg, RipGrepError } from "ripgrep-js";
import {
  findingMatchesFilter,
  DEFAULT_SARIF_FILTER,
  Filter,
  KeyParams,
  findingKey,
  DEFAULT_KEY_PARAMS,
  JSONObject,
  assessmentLink,
} from "./utils";

const DEFAULT_LAB_URL = "https://app.nowsecure.com";

const SARIF_SCHEMA_URL =
  "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json";

/**
 * Convert NowSecure severity to a SARIF notification level.
 */
function severityToNotification(input: string): Notification.level {
  if (input === "high" || input === "critical") {
    return "error";
  } else if (input === "medium" || input === "low") {
    return "warning";
  } else if (input === "info") {
    return "note";
  } else {
    // NOTE: In practice, this should never happen.
    return "none";
  }
}

function severityToScore(input: Severity): number {
  switch (input) {
    case "critical":
      return 9.5;
    case "high":
      return 8.0;
    case "medium":
      return 5.5;
    case "low":
      return 2.0;
    case "info":
      return 0;
  }
}

/**
 * Convert a Platform report to Sarif.
 *
 * The data **must** be provided from the GraphQL query in the root of the repository,
 * otherwise data may be missing, in which case the behavior of this function
 * is undefined.
 */
export async function convertToSarif(
  data: PullReportResponse,
  keyParams: KeyParams = DEFAULT_KEY_PARAMS,
  filter: Filter = DEFAULT_SARIF_FILTER,
  labUrl: string = DEFAULT_LAB_URL,
  rainier: boolean = true
): Promise<Log> {
  const assessment = data.data.auto.assessments[0];
  const report = assessment.report;

  if (!report) {
    throw new Error("No report data");
  }

  report.findings = report.findings.filter((finding) =>
    findingMatchesFilter(finding, filter)
  );

  const rules: ReportingDescriptor[] = [];
  for (const finding of report.findings) {
    if (!finding.affected) continue;

    let markdown = "";
    let issueSummary = "No issue description available.\n";
    if (finding.summary) {
      issueSummary = finding.summary;
    }
    const tags: string[] = [];

    const issue = finding.check.issue;
    if (issue) {
      if (issue.category) {
        tags.push(issue.category);
      }

      if (issue.cvss) {
        tags.push(`CVSS-${issue.cvss.toFixed(2)}`);
      }

      if (issue.impactSummary) {
        markdown += `## Business Impact\n${issue.impactSummary}\n`;
      }

      if (issue.stepsToReproduce) {
        markdown += `## Steps to Reproduce\n${issue.stepsToReproduce}\n`;
      }

      if (issue.recommendation || issue.codeSamples || issue.guidanceLinks) {
        markdown += "## Remediation Resources\n";
      }

      if (issue.recommendation) {
        markdown += `### Recommended Fix\n${issue.recommendation}\n`;
      }

      if (issue.codeSamples && issue.codeSamples.length !== 0) {
        markdown += "## Code Samples\n";

        for (const codeSample of issue.codeSamples) {
          markdown += "<details>\n";
          markdown += `<summary>${codeSample.caption} (click to expand)</summary>\n\n`;
          markdown += "```";
          if (codeSample.syntax) {
            markdown += codeSample.syntax;
          }
          markdown += "\n";
          markdown += codeSample.block;
          markdown += "\n```\n";
          markdown += "</details>\n";
        }
      }

      markdown += "\n";

      if (issue.guidanceLinks && issue.guidanceLinks.length !== 0) {
        markdown += "## Additional Guidance\n";
        for (const guidanceLink of issue.guidanceLinks) {
          markdown += `* ${guidanceLink.caption} ${guidanceLink.url}\n`;
        }
      }

      markdown += "\n";
    }

    // Format NowSecure evidence in a Markdown table, when possible.

    const context = finding.context;
    if (context && context.rows && context.rows.length > 0) {
      // Markdown tables look like:
      // |A|B|C|
      let markdownHeader = "|";
      // |-|-|-|
      let markdownDivider = "|";
      // If the key for "A" is "a", then it will be the first
      // entry in the ordering list, so on...
      const ordering = [];

      for (const fieldName of Object.keys(context.fields)) {
        const fieldTitle = context.fields[fieldName].title;
        markdownHeader += `${fieldTitle}|`;
        markdownDivider += `-|`;
        ordering.push(fieldName);
      }

      markdown += markdownHeader + "\n";
      markdown += markdownDivider + "\n";

      // For each row, insert keys in the order specified in "ordering".
      for (const row of context.rows) {
        markdown += "|";
        for (const key of ordering) {
          let data = row[key];
          if (typeof data !== "string") {
            data = JSON.stringify(data);
          }

          markdown += "```` " + data + " ````" + "|";
        }
        markdown += "\n";
      }
    }

    rules.push({
      id: findingKey(assessment, finding, keyParams),
      name: finding.title,
      helpUri: assessmentLink(labUrl, rainier, assessment, finding),
      shortDescription: {
        text: finding.title,
      },
      fullDescription: {
        text: issueSummary,
      },
      properties: {
        problem: {
          severity: finding.severity,
        },
        tags,
        precision: "medium",
        // security-severity is a string (even though it's a number)
        "security-severity": severityToScore(finding.severity).toString(),
      },
      help: {
        // NOTE: In practice this should not display on the GitHub UI.
        text: "NowSecure only provides recommendations in a Markdown format.",
        markdown,
      },
    });
  }

  const results: Result[] = [];
  for (const finding of report.findings) {
    if (!finding.affected) continue;

    const issue = finding.check.issue;

    let issueDescription = "No issue description available\n";
    if (issue && issue.description) {
      issueDescription = issue.description;
    }

    const level = severityToNotification(finding.severity);
    // If we are missing a specialized result for a rule, show the "simple"
    // result that does not show detailed line number information (refer to the
    // evidence table instead).
    const simpleResult = {
      ruleId: findingKey(assessment, finding, keyParams),
      message: {
        // Markdown doesn't work here. We render our information in the "help"
        // field in the reporting descriptor.
        text: issueDescription,
      },
      level: level,
      locations: [
        {
          // We don't have line number information so instead produce phony
          // information for a file that does not exist.
          physicalLocation: {
            artifactLocation: {
              uri: "Unknown",
              uriBaseId: "%SRCROOT%",
            },
            region: {
              startLine: 1,
              endLine: 1,
              byteOffset: 0,
              byteLength: 0,
            },
          },
        },
      ],
    };

    // FIXME: This should be refactored out and abstracted.
    const localResults = [];

    if (finding.key == "path_traversal") {
      const context = finding.context;
      if (context && context.rows && context.rows.length > 0) {
        for (const row of context.rows) {
          const name = row.name as string;
          const namespace = name.split(".");
          if (namespace.length > 0) {
            const [className] = namespace.slice(-1);
            // NOTE: Source mapping works for Java and Kotlin ONLY.
            const locations = await searchLocations(`class ${className}`);

            for (const physicalLocation of locations) {
              localResults.push({
                ruleId: findingKey(assessment, finding, keyParams),
                message: {
                  text: issueDescription,
                },
                level: level,
                locations: [
                  {
                    physicalLocation,
                  },
                ],
              });
            }
          }
        }
      }
    }

    if (localResults.length !== 0) {
      results.push(...localResults);
    } else {
      results.push(simpleResult);
    }
  }

  const run: Run = {
    tool: {
      driver: {
        name: "NowSecure",
        informationUri: "https://www.nowsecure.com/",
        semanticVersion: "1.0.0",
        rules,
      },
    },
    results,
  };

  const log: Log = {
    $schema: SARIF_SCHEMA_URL,
    version: "2.1.0",
    runs: [run],
  };

  return log;
}

function strip(name: string): string {
  return name.replace(/[^0-9a-z-_\s]/gi, "");
}

/*
 * Search the codebase for a string and return a physical location that corresponds
 * to it.
 */
async function searchLocations(name: string): Promise<PhysicalLocation[]> {
  let results;
  try {
    results = await rg("./", `"${strip(name)}"`);
  } catch (e) {
    const error = e as RipGrepError;
    console.log(`Error: ripgrep: ${error.message}: ${error.stderr}`);
    console.log(
      "Note: ripgrep is required for line-number identification. On Ubuntu-based distributions use `apt-get install -y ripgrep` before running the NowSecure action"
    );
    return [];
  }

  const locations = [];

  for (const result of results) {
    locations.push({
      artifactLocation: {
        uri: result.path.text,
        uriBaseId: "%SRCROOT%",
      },
      region: {
        startLine: result.line_number,
        endLine: result.line_number,
        byteOffset: 0,
        byteLength: 0,
      },
    });
  }

  return locations;
}
