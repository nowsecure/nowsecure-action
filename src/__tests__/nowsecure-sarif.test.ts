/*
 * Copyright Â© 2022 NowSecure Inc.
 *
 * SPDX-License-Identifier: MIT
 */

/* eslint-disable @typescript-eslint/no-unused-vars */

import { convertToSarif } from "../nowsecure-sarif";
import { NowSecure, DEFAULT_API_URL } from "../nowsecure-client";
import nock from "nock";
import path from "path";
import { HttpClientError } from "@actions/http-client";

const platformToken = "AAABBB";
const assessmentId = "CCCDDD";

// Silence warnings from configuration manager.
const consoleWarnMock = jest.spyOn(console, "warn").mockImplementation();

describe("SARIF conversion", () => {
  const ns = new NowSecure(platformToken);

  test("can perform conversion", async () => {
    const scope = nock(DEFAULT_API_URL)
      .get("/graphql")
      .query(true)
      .replyWithFile(
        200,
        path.join(__dirname, "resources", "response_200.json")
      );

    const report = await ns.pullReport(assessmentId);
    const sarif = await convertToSarif(report);
    expect(sarif).toMatchSnapshot();
  });

  test.each([1, 2, 3])(
    "performs retries for 502 errors (%d times)",
    async (times) => {
      const scope1 = nock(DEFAULT_API_URL)
        .get("/graphql")
        .query(true)
        .times(times)
        .reply(502);

      const scope2 = nock(DEFAULT_API_URL)
        .get("/graphql")
        .query(true)
        .replyWithFile(
          200,
          path.join(__dirname, "resources", "response_200.json")
        );

      const report = await ns.pullReport(assessmentId);
      const sarif = await convertToSarif(report);
      expect(sarif).toMatchSnapshot();
      expect(scope1.isDone());
      expect(scope2.isDone());
    }
  );

  test("request fails after 4 consequtive 502 errors", async () => {
    const scope1 = nock(DEFAULT_API_URL)
      .get("/graphql")
      .query(true)
      .times(4)
      .reply(502);

    expect(ns.pullReport(assessmentId)).rejects.toThrow(HttpClientError);
  });
});
