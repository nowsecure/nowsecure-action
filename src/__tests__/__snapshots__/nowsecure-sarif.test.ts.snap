// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SARIF conversion can perform conversion 1`] = `
Object {
  "$schema": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json",
  "runs": Array [
    Object {
      "results": Array [
        Object {
          "level": "warning",
          "locations": Array [
            Object {
              "physicalLocation": Object {
                "artifactLocation": Object {
                  "uri": "Unknown",
                  "uriBaseId": "%SRCROOT%",
                },
                "region": Object {
                  "byteLength": 0,
                  "byteOffset": 0,
                  "endLine": 1,
                  "startLine": 1,
                },
              },
            },
          ],
          "message": Object {
            "text": "This is a descriptive explanation of the example check.",
          },
          "ruleId": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
        },
      ],
      "tool": Object {
        "driver": Object {
          "informationUri": "https://www.nowsecure.com/",
          "name": "NowSecure",
          "rules": Array [
            Object {
              "fullDescription": Object {
                "text": "This check is an example check.",
              },
              "help": Object {
                "markdown": "## Business Impact
This is the impact of the example check.
## Steps to Reproduce
These are steps to reproduce the issue.
## Remediation Resources
### Recommended Fix
This is an example recommendation.


",
                "text": "NowSecure only provides recommendations in a Markdown format.",
              },
              "helpUri": "https://app.nowsecure.com/app/00000000-0000-0000-0000-000000000000/assessment/00000000-0000-0000-0000-000000000001/findings#finding-example_check",
              "id": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
              "name": "This Is An Example Check",
              "properties": Object {
                "precision": "medium",
                "problem": Object {
                  "severity": "medium",
                },
                "security-severity": "5.5",
                "tags": Array [
                  "CVSS-10.00",
                ],
              },
              "shortDescription": Object {
                "text": "This Is An Example Check",
              },
            },
          ],
          "semanticVersion": "1.0.0",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;

exports[`SARIF conversion performs retries for 502 errors (1 times) 1`] = `
Object {
  "$schema": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json",
  "runs": Array [
    Object {
      "results": Array [
        Object {
          "level": "warning",
          "locations": Array [
            Object {
              "physicalLocation": Object {
                "artifactLocation": Object {
                  "uri": "Unknown",
                  "uriBaseId": "%SRCROOT%",
                },
                "region": Object {
                  "byteLength": 0,
                  "byteOffset": 0,
                  "endLine": 1,
                  "startLine": 1,
                },
              },
            },
          ],
          "message": Object {
            "text": "This is a descriptive explanation of the example check.",
          },
          "ruleId": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
        },
      ],
      "tool": Object {
        "driver": Object {
          "informationUri": "https://www.nowsecure.com/",
          "name": "NowSecure",
          "rules": Array [
            Object {
              "fullDescription": Object {
                "text": "This check is an example check.",
              },
              "help": Object {
                "markdown": "## Business Impact
This is the impact of the example check.
## Steps to Reproduce
These are steps to reproduce the issue.
## Remediation Resources
### Recommended Fix
This is an example recommendation.


",
                "text": "NowSecure only provides recommendations in a Markdown format.",
              },
              "helpUri": "https://app.nowsecure.com/app/00000000-0000-0000-0000-000000000000/assessment/00000000-0000-0000-0000-000000000001/findings#finding-example_check",
              "id": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
              "name": "This Is An Example Check",
              "properties": Object {
                "precision": "medium",
                "problem": Object {
                  "severity": "medium",
                },
                "security-severity": "5.5",
                "tags": Array [
                  "CVSS-10.00",
                ],
              },
              "shortDescription": Object {
                "text": "This Is An Example Check",
              },
            },
          ],
          "semanticVersion": "1.0.0",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;

exports[`SARIF conversion performs retries for 502 errors (2 times) 1`] = `
Object {
  "$schema": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json",
  "runs": Array [
    Object {
      "results": Array [
        Object {
          "level": "warning",
          "locations": Array [
            Object {
              "physicalLocation": Object {
                "artifactLocation": Object {
                  "uri": "Unknown",
                  "uriBaseId": "%SRCROOT%",
                },
                "region": Object {
                  "byteLength": 0,
                  "byteOffset": 0,
                  "endLine": 1,
                  "startLine": 1,
                },
              },
            },
          ],
          "message": Object {
            "text": "This is a descriptive explanation of the example check.",
          },
          "ruleId": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
        },
      ],
      "tool": Object {
        "driver": Object {
          "informationUri": "https://www.nowsecure.com/",
          "name": "NowSecure",
          "rules": Array [
            Object {
              "fullDescription": Object {
                "text": "This check is an example check.",
              },
              "help": Object {
                "markdown": "## Business Impact
This is the impact of the example check.
## Steps to Reproduce
These are steps to reproduce the issue.
## Remediation Resources
### Recommended Fix
This is an example recommendation.


",
                "text": "NowSecure only provides recommendations in a Markdown format.",
              },
              "helpUri": "https://app.nowsecure.com/app/00000000-0000-0000-0000-000000000000/assessment/00000000-0000-0000-0000-000000000001/findings#finding-example_check",
              "id": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
              "name": "This Is An Example Check",
              "properties": Object {
                "precision": "medium",
                "problem": Object {
                  "severity": "medium",
                },
                "security-severity": "5.5",
                "tags": Array [
                  "CVSS-10.00",
                ],
              },
              "shortDescription": Object {
                "text": "This Is An Example Check",
              },
            },
          ],
          "semanticVersion": "1.0.0",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;

exports[`SARIF conversion performs retries for 502 errors (3 times) 1`] = `
Object {
  "$schema": "https://raw.githubusercontent.com/schemastore/schemastore/master/src/schemas/json/sarif-2.1.0-rtm.5.json",
  "runs": Array [
    Object {
      "results": Array [
        Object {
          "level": "warning",
          "locations": Array [
            Object {
              "physicalLocation": Object {
                "artifactLocation": Object {
                  "uri": "Unknown",
                  "uriBaseId": "%SRCROOT%",
                },
                "region": Object {
                  "byteLength": 0,
                  "byteOffset": 0,
                  "endLine": 1,
                  "startLine": 1,
                },
              },
            },
          ],
          "message": Object {
            "text": "This is a descriptive explanation of the example check.",
          },
          "ruleId": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
        },
      ],
      "tool": Object {
        "driver": Object {
          "informationUri": "https://www.nowsecure.com/",
          "name": "NowSecure",
          "rules": Array [
            Object {
              "fullDescription": Object {
                "text": "This check is an example check.",
              },
              "help": Object {
                "markdown": "## Business Impact
This is the impact of the example check.
## Steps to Reproduce
These are steps to reproduce the issue.
## Remediation Resources
### Recommended Fix
This is an example recommendation.


",
                "text": "NowSecure only provides recommendations in a Markdown format.",
              },
              "helpUri": "https://app.nowsecure.com/app/00000000-0000-0000-0000-000000000000/assessment/00000000-0000-0000-0000-000000000001/findings#finding-example_check",
              "id": "5e63241d7f766e1697c5a450f8178a53bdcc66af0d5bcc73775749a044896c7e",
              "name": "This Is An Example Check",
              "properties": Object {
                "precision": "medium",
                "problem": Object {
                  "severity": "medium",
                },
                "security-severity": "5.5",
                "tags": Array [
                  "CVSS-10.00",
                ],
              },
              "shortDescription": Object {
                "text": "This Is An Example Check",
              },
            },
          ],
          "semanticVersion": "1.0.0",
        },
      },
    },
  ],
  "version": "2.1.0",
}
`;
